README.md
# SenderBot

SenderBot is a Telegram Bot that sends messages to Groups and Channels of Telegram.

For more information, please visit our website:

https://www.atareao.es/

# Requirements

Required dependencies:

```
    python3,
    python3-requests
    python3-future
    python3-certifi
```
# Install

To install SenderBot in Ubuntu, Linux Mint and derivatives, run these commands in a terminal,

```
sudo add-apt-repository ppa:atareao/telegram
sudo apt update
sudo apt install senderbot
```


debian/changelog
senderbot (0.4.0-0extras17.10.0) artful; urgency=medium

  * Inline bot

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 26 Oct 2017 13:46:51 +0200

senderbot (0.3.4-0extras17.04.0) zesty; urgency=medium

  * Picamera
  * Send images from url

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 18 Oct 2017 14:57:32 +0200

senderbot (0.3.3-0extras17.04.4) zesty; urgency=medium

  * Added OpenVPN status

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 13 Oct 2017 13:41:25 +0200

senderbot (0.3.2-0extras17.04.0) zesty; urgency=medium

  * Added servicios

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 30 Sep 2017 08:27:37 +0200

senderbot (0.3.1-0extras17.04.0) zesty; urgency=medium

  * Working alone
  * Looking for services

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 30 Sep 2017 08:07:05 +0200

senderbot (0.2.1-0extras17.04.0) zesty; urgency=medium

  * Added dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 21 Sep 2017 12:49:21 +0200

senderbot (0.2.0-0extras17.04.0) zesty; urgency=medium

  * Added services to send message on start o shutdown

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 21 Sep 2017 12:38:18 +0200

senderbot (0.1.0-0extras17.04.4) zesty; urgency=medium

  * First release

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 09 Sep 2017 19:13:43 +0200


raebot.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# raebot.py
#
# This file is part of ---
#
# Copyright (C) 2017
# Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from telegram import InlineQueryResultArticle, InputTextMessageContent
from telegram.ext import InlineQueryHandler


def inline_caps(bot, update):
     query = update.inline_query.query
     if not query:
         return
     results = list()
     results.append(
         InlineQueryResultArticle(
             id=query.upper(),
             title='Caps',
             input_message_content=InputTextMessageContent(query.upper())
         )
     )
     bot.answer_inline_query(update.inline_query.id, results)


inline_caps_handler = InlineQueryHandler(inline_caps)
dispatcher.add_handler(inline_caps_handler)

src/senderbot2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Basic example for a bot that uses inline keyboards.
# This program is dedicated to the public domain under the CC0 license.

import pprint
import json
import codecs
import os
import logging

from telegram import InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Updater, CommandHandler, CallbackQueryHandler


APP = 'senderbot'
APPNAME = 'SenderBot'
APPCONF = APP + '.conf'
CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config')
CONFIG_APP_DIR = os.path.join(CONFIG_DIR, APP)
CONFIG_FILE = os.path.join(CONFIG_APP_DIR, APPCONF)
PARAMS = {'token': None,
          'channel_id': None}

logging.basicConfig(format='%(asctime)s - %(name)s - %(message)s',
                    filename='/tmp/senderbot.log',
                    filemode='w',
                    level=logging.INFO)


class Configuration(object):
    def __init__(self):
        self.params = PARAMS
        self.read()

    def get(self, key):
        try:
            return self.params[key]
        except KeyError as e:
            print(e)
            self.params[key] = PARAMS[key]
            return self.params[key]

    def set(self, key, value):
        self.params[key] = value

    def reset(self):
        if os.path.exists(CONFIG_FILE):
            os.remove(CONFIG_FILE)
        self.params = PARAMS
        self.save()

    def set_defaults(self):
        self.params = PARAMS
        self.save()

    def print_config(self):
        pprint(self.params)

    def read(self):
        try:
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        except IOError as e:
            print(e)
            self.save()
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        try:
            self.params = json.loads(f.read())
        except ValueError as e:
            print(e)
            self.save()
        f.close()

    def save(self):
        f = codecs.open(CONFIG_FILE, 'w', 'utf-8')
        f.write(json.dumps(self.params))
        f.close()


def start(bot, update):
    keyboard = [[InlineKeyboardButton("Option 1", callback_data='1'),
                 InlineKeyboardButton("Option 2", callback_data='2')],

                [InlineKeyboardButton("Option 3", callback_data='3')]]

    reply_markup = InlineKeyboardMarkup(keyboard)

    update.message.reply_text('Please choose:', reply_markup=reply_markup)


def button(bot, update):
    query = update.callback_query

    bot.edit_message_text(text="Selected option: %s" % query.data,
                          chat_id=query.message.chat_id,
                          message_id=query.message.message_id)


def help(bot, update):
    update.message.reply_text("Use /start to test this bot.")


def error(bot, update, error):
    logging.warning('Update "%s" caused error "%s"' % (update, error))


configuration = Configuration()
token = configuration.get('token')


# Create the Updater and pass it your bot's token.
updater = Updater(token)

updater.dispatcher.add_handler(CommandHandler('start', start))
updater.dispatcher.add_handler(CallbackQueryHandler(button))
updater.dispatcher.add_handler(CommandHandler('help', help))
updater.dispatcher.add_error_handler(error)

# Start the Bot
updater.start_polling()

# Run the bot until the user presses Ctrl-C or the process receives SIGINT,
# SIGTERM or SIGABRT
updater.idle()


src/senderbot3.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Simple Bot to reply to Telegram messages
# This program is dedicated to the public domain under the CC0 license.
"""
This Bot uses the Updater class to handle the bot.

First, a few callback functions are defined. Then, those functions are passed to
the Dispatcher and registered at their respective places.
Then, the bot is started and runs until we press Ctrl-C on the command line.

Usage:
Example of a bot-user conversation using ConversationHandler.
Send /start to initiate the conversation.
Press Ctrl-C on the command line or send a signal to the process to stop the
bot.
"""

import pprint
import json
import codecs
import os
import logging

from telegram import (ReplyKeyboardMarkup, ReplyKeyboardRemove)
from telegram.ext import (Updater, CommandHandler, MessageHandler, Filters,
                          RegexHandler, ConversationHandler)

logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - \
%(message)s',
                    level=logging.INFO,
                    filename='/tmp/senderbot.log',
                    filemode='w')
logger = logging.getLogger(__name__)

APP = 'senderbot'
APPNAME = 'SenderBot'
APPCONF = APP + '.conf'
CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config')
CONFIG_APP_DIR = os.path.join(CONFIG_DIR, APP)
CONFIG_FILE = os.path.join(CONFIG_APP_DIR, APPCONF)
PARAMS = {'token': None,
          'channel_id': None}

if not os.path.exists(CONFIG_APP_DIR):
    os.makedirs(CONFIG_APP_DIR)


class Configuration(object):
    def __init__(self):
        self.params = PARAMS
        self.read()

    def get(self, key):
        try:
            return self.params[key]
        except KeyError as e:
            print(e)
            self.params[key] = PARAMS[key]
            return self.params[key]

    def set(self, key, value):
        self.params[key] = value

    def reset(self):
        if os.path.exists(CONFIG_FILE):
            os.remove(CONFIG_FILE)
        self.params = PARAMS
        self.save()

    def set_defaults(self):
        self.params = PARAMS
        self.save()

    def print_config(self):
        pprint(self.params)

    def read(self):
        try:
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        except IOError as e:
            print(e)
            self.save()
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        try:
            self.params = json.loads(f.read())
        except ValueError as e:
            print(e)
            self.save()
        f.close()

    def save(self):
        f = codecs.open(CONFIG_FILE, 'w', 'utf-8')
        f.write(json.dumps(self.params))
        f.close()
# Enable logging

GENDER, PHOTO, LOCATION, BIO = range(4)


def start(bot, update):
    reply_keyboard = [['Boy', 'Girl', 'Other']]

    update.message.reply_text(
        'Hi! My name is Professor Bot. I will hold a conversation with you. '
        'Send /cancel to stop talking to me.\n\n'
        'Are you a boy or a girl?',
        reply_markup=ReplyKeyboardMarkup(reply_keyboard, one_time_keyboard=True))

    return GENDER


def gender(bot, update):
    user = update.message.from_user
    logger.info("Gender of %s: %s" % (user.first_name, update.message.text))
    update.message.reply_text('I see! Please send me a photo of yourself, '
                              'so I know what you look like, or send /skip if you don\'t want to.',
                              reply_markup=ReplyKeyboardRemove())

    return PHOTO


def photo(bot, update):
    user = update.message.from_user
    photo_file = bot.get_file(update.message.photo[-1].file_id)
    photo_file.download('user_photo.jpg')
    logger.info("Photo of %s: %s" % (user.first_name, 'user_photo.jpg'))
    update.message.reply_text('Gorgeous! Now, send me your location please, '
                              'or send /skip if you don\'t want to.')

    return LOCATION


def skip_photo(bot, update):
    user = update.message.from_user
    logger.info("User %s did not send a photo." % user.first_name)
    update.message.reply_text('I bet you look great! Now, send me your location please, '
                              'or send /skip.')

    return LOCATION


def location(bot, update):
    user = update.message.from_user
    user_location = update.message.location
    logger.info("Location of %s: %f / %f"
                % (user.first_name, user_location.latitude, user_location.longitude))
    update.message.reply_text('Maybe I can visit you sometime! '
                              'At last, tell me something about yourself.')

    return BIO


def skip_location(bot, update):
    user = update.message.from_user
    logger.info("User %s did not send a location." % user.first_name)
    update.message.reply_text('You seem a bit paranoid! '
                              'At last, tell me something about yourself.')

    return BIO


def bio(bot, update):
    user = update.message.from_user
    logger.info("Bio of %s: %s" % (user.first_name, update.message.text))
    update.message.reply_text('Thank you! I hope we can talk again some day.')

    return ConversationHandler.END


def cancel(bot, update):
    user = update.message.from_user
    logger.info("User %s canceled the conversation." % user.first_name)
    update.message.reply_text('Bye! I hope we can talk again some day.',
                              reply_markup=ReplyKeyboardRemove())

    return ConversationHandler.END


def error(bot, update, error):
    logger.warn('Update "%s" caused error "%s"' % (update, error))


def main():
    # Create the EventHandler and pass it your bot's token.
    configuration = Configuration()
    token = configuration.get('token')
    updater = Updater(token)

    # Get the dispatcher to register handlers
    dp = updater.dispatcher

    # Add conversation handler with the states GENDER, PHOTO, LOCATION and BIO
    conv_handler = ConversationHandler(
        entry_points=[CommandHandler('start', start)],

        states={
            GENDER: [RegexHandler('^(Boy|Girl|Other)$', gender)],

            PHOTO: [MessageHandler(Filters.photo, photo),
                    CommandHandler('skip', skip_photo)],

            LOCATION: [MessageHandler(Filters.location, location),
                       CommandHandler('skip', skip_location)],

            BIO: [MessageHandler(Filters.text, bio)]
        },

        fallbacks=[CommandHandler('cancel', cancel)]
    )

    dp.add_handler(conv_handler)

    # log all errors
    dp.add_error_handler(error)

    # Start the Bot
    updater.start_polling()

    # Run the bot until you press Ctrl-C or the process receives SIGINT,
    # SIGTERM or SIGABRT. This should be used most of the time, since
    # start_polling() is non-blocking and will stop the bot gracefully.
    updater.idle()


if __name__ == '__main__':
    main()


src/senderbot4.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Basic example for a bot that uses inline keyboards.
# This program is dedicated to the public domain under the CC0 license.

# --- Start Configuration dependencies
import os
import json
import codecs
# --- End Configuration dependencies
import logging

from telegram import InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Updater, CommandHandler, CallbackQueryHandler

APP = 'senderbot'
APPNAME = 'SenderBot'
APPCONF = APP + '.conf'
CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config')
CONFIG_APP_DIR = os.path.join(CONFIG_DIR, APP)
CONFIG_FILE = os.path.join(CONFIG_APP_DIR, APPCONF)
PARAMS = {'token': None,
          'channel_id': None}

logging.basicConfig(format='%(asctime)s - %(name)s - %(message)s',
                    filename='/tmp/senderbot.log',
                    filemode='w',
                    level=logging.INFO)


class Configuration(object):
    def __init__(self):
        self.params = PARAMS
        self.read()

    def get(self, key):
        try:
            return self.params[key]
        except KeyError as e:
            print(e)
            self.params[key] = PARAMS[key]
            return self.params[key]

    def set(self, key, value):
        self.params[key] = value

    def reset(self):
        if os.path.exists(CONFIG_FILE):
            os.remove(CONFIG_FILE)
        self.params = PARAMS
        self.save()

    def set_defaults(self):
        self.params = PARAMS
        self.save()

    def read(self):
        try:
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        except IOError as e:
            print(e)
            self.save()
            f = codecs.open(CONFIG_FILE, 'r', 'utf-8')
        try:
            self.params = json.loads(f.read())
        except ValueError as e:
            print(e)
            self.save()
        f.close()

    def save(self):
        f = codecs.open(CONFIG_FILE, 'w', 'utf-8')
        f.write(json.dumps(self.params))
        f.close()


def start(bot, update):
    keyboard = [[InlineKeyboardButton("Option 1", callback_data='1'),
                 InlineKeyboardButton("Option 2", callback_data='2')],

                [InlineKeyboardButton("Option 3", callback_data='3')]]

    reply_markup = InlineKeyboardMarkup(keyboard)

    update.message.reply_text('Please choose:', reply_markup=reply_markup)


def button(bot, update):
    query = update.callback_query

    bot.edit_message_text(text="Selected option: %s" % query.data,
                          chat_id=query.message.chat_id,
                          message_id=query.message.message_id)


def help(bot, update):
    update.message.reply_text("Use /start to test this bot.")


def error(bot, update, error):
    logging.warning('Update "%s" caused error "%s"' % (update, error))


configuration = Configuration()
token = configuration.get('token')

# Create the Updater and pass it your bot's token.
updater = Updater(token)

updater.dispatcher.add_handler(CommandHandler('start', start))
updater.dispatcher.add_handler(CallbackQueryHandler(button))
updater.dispatcher.add_handler(CommandHandler('help', help))
updater.dispatcher.add_error_handler(error)

# Start the Bot
updater.start_polling()#allowed_updates=[])

# Run the bot until the user presses Ctrl-C or the process receives SIGINT,
# SIGTERM or SIGABRT
updater.idle()


src/rae.py
#!/usr/bin/python3

import ctypes
import urllib.parse
import bs4
import requests

s = requests.Session()


class _Shared:
    PARSER = 'lxml'

    def __init__(self):
        pass

    @staticmethod
    def solve_challenge(c, slt, s1, s2, n, table):
        m = pow(ord(s2) - ord(s1) + 1, n)
        arr = [s1] * n
        chlg = None

        for _i in range(m-1):
            for j in range(n-1, -1, -1):
                arr[j] = chr(ord(arr[j]) + 1)

                if arr[j] <= s2:
                    break

                arr[j] = s1

            chlg = ''.join(arr)
            crc = -1

            for k in chlg + slt:
                index = ((crc ^ ord(k)) & 0x000000FF) * 9
                x = int(table[index:index+8], 16)
                crc = ctypes.c_int32(crc >> 8).value ^ ctypes.c_int32(x).value

            crc = abs(crc ^ -1)

            if crc == c:
                break

        return chlg

    @staticmethod
    def get_payload(r, rf):
        try:
            tmp = r.index('document.forms[0].elements[1].value=\"') + 37
            first = r[tmp:r.index(':', tmp)]

            tmp = r.index('var slt = \"') + 11
            slt = r[tmp:r.index('\"', tmp)]

            tmp = r.index('var c = ') + 8
            c = int(r[tmp:r.index('\r', tmp)])

            tmp = r.index('var s1 = \'') + 10
            s1 = r[tmp:r.index('\'', tmp)]

            tmp = r.index('var s2 = \'') + 10
            s2 = r[tmp:r.index('\'', tmp)]

            tmp = r.index('var n = ') + 8
            n = int(r[tmp:r.index('\n', tmp)])

            tmp = r.index('var table = \"') + 13
            table = r[tmp:r.index('\"', tmp)]
        except ValueError:
            return None

        chlg = _Shared.solve_challenge(c, slt, s1, s2, n, table)

        if chlg is None:
            return None

        cr = ':'.join([first, chlg, slt, str(c)])

        return [['TS017111a7_id', '3'],
                ['TS017111a7_cr', cr],
                ['TS017111a7_76', '0'],
                ['TS017111a7_86', '0'],
                ['TS017111a7_md', '1'],
                ['TS017111a7_rf', rf],
                ['TS017111a7_ct', '0'],
                ['TS017111a7_pd', '0']]

    @staticmethod
    def do_request(request_url, rf, do_post=True):
        r = s.get(request_url)

        if r.status_code != requests.codes.ok:
            return None

        if do_post:
            payload = _Shared.get_payload(r.text, rf)

            if payload is None:
                return None

            r = s.post(request_url, data=payload)

        if r.status_code == requests.codes.ok:
            return r


class DLE:
    HOST = 'http://dle.rae.es'
    URL_RANDOM_WORD = HOST + '/srv/random'
    URL_TODAYS_WORD = HOST + '/srv/wotd'
    URL_FETCH = HOST + '/srv/fetch'
    URL_SEARCH = HOST + '/srv/search'
    MAX_LEMMAS_PAGE = 200

    def __init__(self):
        pass

    @staticmethod
    def _conjugate(name, data, col):
        result = [name]

        for i in data:
            result.append([i[0], i[col]])

        return result

    @staticmethod
    def conjugate_verb(verb):
        verb_id = DLE.search_word(verb)[1]

        return DLE.conjugate_id(verb_id)

    @staticmethod
    def conjugate_id(verb_id):
        r = _Shared.do_request(DLE.URL_FETCH + '?id=' + verb_id, 'http://www.rae.es/')

        if r.status_code != requests.codes.ok:
            return None

        soup = bs4.BeautifulSoup(r.text, _Shared.PARSER)
        cnj = soup.find('table', class_='cnj')

        if cnj is None:
            return None

        data = []
        h = []

        for row in cnj.find_all('tr'):
            cells = [cell.text.strip() for cell in row.find_all('td')]
            data.append([cell for cell in cells if cell])
            heads = [header.text.strip() for header in row.find_all('th')]
            h.append([header for header in heads if header])

        data = [e for e in data if e]

        vc = [h[0][0]]
        vc.append([h[1][0], data[0][0]])
        vc.append([h[1][1], data[0][1]])
        vc.append([h[3][0], data[1][0]])
        vc.append(DLE._conjugate(h[5][0] + ' ' + h[6][3], data[2:10], 1))
        vc.append(DLE._conjugate(h[5][0] + ' ' + h[6][4], data[2:10], 2))
        vc.append(DLE._conjugate(h[5][0] + ' ' + h[15][0], data[10:18], 1))
        vc.append(DLE._conjugate(h[5][0] + ' ' + h[15][1], data[10:18], 2))
        vc.append(DLE._conjugate(h[5][0] + ' ' + h[24][0], data[18:26], 1))
        vc.append(DLE._conjugate(h[33][0] + ' ' + h[34][3], data[26:34], 1))
        vc.append(DLE._conjugate(h[33][0] + ' ' + h[34][4], data[26:34], 2))
        vc.append(DLE._conjugate(h[33][0] + ' ' + h[43][0], data[34:42], 1))
        vc.append(h[52][0])
        vc.append([data[42][0], data[42][1]])
        vc.append([data[43][0], data[43][1]])
        vc.append([data[44][0], data[44][1]])
        vc.append([data[45][0], data[45][1]])

        return vc

    @staticmethod
    def random_word():
        s.cookies.clear()

        r = _Shared.do_request(DLE.URL_RANDOM_WORD, DLE.URL_RANDOM_WORD, False)

        if r is None:
            return None

        soup = bs4.BeautifulSoup(r.text, _Shared.PARSER)
        article = soup.article

        if article is not None:
            return article.text

    @staticmethod
    def _request_word(word, after_host, m=None):
        url = DLE.HOST + '/?w=' + word
        url2 = DLE.HOST + after_host + word

        if m is not None:
            url += m
            url2 += m

        if _Shared.do_request(url, 'http://www.rae.es/') is None:
            return None

        r = _Shared.do_request(url2, url2)

        if r is None:
            return None

        return bs4.BeautifulSoup(r.text, _Shared.PARSER)

    @staticmethod
    def _options(soup):
        results = []

        for op in soup.find_all('a'):
            words = op.text.split('; ')
            word_ids = op.get('href').replace('fetch?id=', '').split('|')

            for word, word_id in zip(words, word_ids):
                results.append([word, word_id])

        return results

    @staticmethod
    def search_id(word_id):
        payload = {'id': word_id}
        r = s.get(DLE.URL_FETCH, data=payload)

        if r.status_code != requests.codes.ok:
            return None

        soup = bs4.BeautifulSoup(r.text, _Shared.PARSER)
        article = soup.article

        if article is not None:
            return article.text

    @staticmethod
    def search_word(word, m=None):
        s.cookies.clear()

        soup = DLE._request_word(word, '/srv/search?w=', m)

        if soup is None:
            return None

        f0 = soup.find('div', id_='f0')

        verb_id = None

        if f0 is not None:
            result = f0.span.text
        else:
            article = soup.article
            result = article.text if article is not None else DLE._options(soup)

            if result is None:
                return None

            e2 = soup.find('a', class_='e2')

            if e2 is not None:
                verb_id = e2['href'].replace('fetch?id=', '')

        return result if isinstance(result, list) else [result, verb_id]

    @staticmethod
    def exact(word):
        return DLE.search_word(word, '&m=30')

    @staticmethod
    def starts_with(prefix):
        return DLE.search_word(prefix, '&m=31')

    @staticmethod
    def ends_with(suffix):
        return DLE.search_word(suffix, '&m=32')

    @staticmethod
    def contains(substring):
        return DLE.search_word(substring, '&m=33')

    @staticmethod
    def anagrams(word):
        s.cookies.clear()

        soup = DLE._request_word(word, '/srv/anagram?w=')

        if soup is not None:
            return DLE._options(soup)

    @staticmethod
    def todays_word():
        s.cookies.clear()

        if _Shared.do_request(DLE.HOST + '/?w=', 'http://www.rae.es/') is None:
            return None

        payload = {'': 'diccionario'}
        r = s.get(DLE.URL_SEARCH, data=payload)

        if r.status_code != requests.codes.ok:
            return None

        r = s.get(DLE.URL_TODAYS_WORD)

        if r.status_code != requests.codes.ok:
            return None

        soup = bs4.BeautifulSoup(r.text, _Shared.PARSER)
        word = soup.a
        word_id = word['href'].replace('id=', '')

        return [word.text, word_id]

    @staticmethod
    def get_lemmas():
        letters = 'aábcdeéfghiíjklmnñoópqrstuúvwxyz'

        prefix = letters[0]
        result = []

        while prefix:
            tmp = DLE.starts_with(prefix)
            nxt = True

            if tmp is not None and tmp:
                current = [i[0] for i in tmp if i is not None]

                if len(current) < DLE.MAX_LEMMAS_PAGE:
                    result.extend(list(set(current) - set(result)))
                else:
                    prefix += letters[0]
                    nxt = False

            if nxt:
                if prefix[-1] == letters[-1]:
                    prefix = prefix[:-1]

                if prefix != letters[-1]:
                    i = letters.find(prefix[-1])
                    prefix = prefix[:-1] + letters[i + 1]

            if prefix == letters[-1]:
                prefix = ''

        result.sort()

        return result


class DPD:
    URL_SEARCH = 'http://lema.rae.es/dpd/?key='

    def __init__(self):
        pass

    @staticmethod
    def search(word):
        s.cookies.clear()

        w = urllib.parse.quote_plus(word)
        r = _Shared.do_request(DPD.URL_SEARCH + w, DPD.URL_SEARCH + w)

        if r is None:
            return None

        soup = bs4.BeautifulSoup(r.text, _Shared.PARSER)
        article = soup.article

        if article is not None:
            return article.text


class DEJ:
    def __init__(self):
        pass


def main():
    pass

if __name__ == '__main__':
    dpd = DLE()
    print(dpd.exact('casa'))
    